src/java.base/share/classes/java/security/Provider.java

```diff
@@ -684,30 +683,737 @@ private void checkInitialized() {
+                        continue;
+                    }
+                }
+                if (!keysToBeKept.contains(aliasKey)) {
+                    services.remove(aliasKey);
+                }
+
+                // Update the Properties map to reflect the alias removal. Note:
+                // oldPropKey cannot be null because aliases always have a
+                // corresponding Properties map entry.
+                oldPropKey = serviceProps.remove(aliasKey);
+                assert oldPropKey != null :
+                        "Unexpected null Property value for an alias.";
+                Provider.super.remove(oldPropKey);
+            }
+
+            // Remove registered service attributes.
+            Map<UString, String> oldAttrProps =
+                    serviceAttrProps.remove(mi.algKey);
+            if (oldAttrProps != null) {
+                for (String oldAttrPropKey : oldAttrProps.values()) {
+                    // Update the Properties map to reflect the attribute
+                    // removal. Note: oldAttrPropKey cannot be null because
+                    // attributes always have a corresponding Properties map
+                    // entry.
+                    assert oldAttrPropKey != null :
+                            "Unexpected null Property value for an attribute.";
+                    Provider.super.remove(oldAttrPropKey);
+                }
+            }
+
+            notifyChanges();
+
+            Provider.this.checkAndUpdateSecureRandom(mi.svc.algKey, false);
+
+            return true;
+        }
+
+        /*
+         * Methods to add, modify and remove services on the map according to
+         * the Legacy API. These methods update the Properties map to reflect
+         * service changes, including algorithms, aliases and attributes.
+         *
+         * Services added with the Current API cannot be overwritten with this
+         * API.
+         *
+         * Notice that these methods do not guarantee atomicity in a sequence of
+         * operations from a service reader point of view. As an example, a
+         * service reader may get a service missing an attribute if looked up
+         * between a writer's ServicesMap::putClassNameLegacy and
+         * ServicesMap::putAttributeLegacy calls. For atomic changes with the
+         * Legacy API see Provider::putAll.
+         */
+
+        boolean putClassNameLegacy(ServiceKey key, String className,
+                String propKey) {
+            assert key != null && className != null && propKey != null :
+                    "Service information missing.";
+            return updateSvc(key, (MappingInfo oldMi, Service newSvc) -> {
+                String canonicalPropKey = propKey;
+                if (oldMi.svc != null) {
+                    // The service exists. Get its Properties map entry. Note:
+                    // Services added through an alias or an attribute may don't
```

Comment:
"may don't"? Maybe you mean "may not" or simply "don't"

Reply:
Good catch. I'd go with "may not" because they have have been added without and later obtained one.